README.md
# CaPredictor
![Website](CaPredictor.png)
- A Flask App that runs a machine learning model for NHL Contracts and predicts a valuation for 
the player in question via the player's stats from the previous season
- [Visit CaPredictor here!](https://capredictor.herokuapp.com)
- Since I'm using a Free Heroku account, sometimes the app goes in hibernation and takes some 
time to open back up. Do not fret - Just reload and let it take its time!

## Motivation
- I have enjoyed hockey for a very long time and as a born-and-raised Canadian, hockey flows through my veins. 
  As an Avid Fan of the Toronto Maple Leafs, I was not impressed by some of the contracts being handed out by
  the team's management, and I wanted to use my skills to create a valuator for NHL players based on their stats.
- I enlisted the help of [my brother](https://github.com/DanyalZaidiPythonGithub), and together we collected data 
  for various NHL players to make models for contracts in the NHL

### Getting Started
- Visit the website at https://capredictor.herokuapp.com.
- When running a program of this sort on your own machine, you will need to have several python files installed, preferably
within a virtual environment and the required files can be seen in the `[Requirements File](requirements.txt)`
- Knowledge of Front-End development and Back-End development(through Flask) is required. Additionally, 
Knowledge of Python 3 is required

#### Installation
- I edited the app in both VScode and the Pycharm IDE, testing the program through my local machine
- To run the flask app, create a file named `app.py` and import the modules present in this [file](app.py)
- Add the following code under the file imports:
\
`app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI']='sqlite:///requests.db'
db = SQLAlchemy(app)`

- Make sure to add the following code at the bottom of the file:
\
`if __name__ == '__main__':
	app.run(debug=False)`
- Create a database in the working directory by running the following code in a terminal:
\
`python
from app import db
db.create_all()`

- To delete the database, run the following in a terminal:
\
`python
from app import db
db.delete_all()`

- This will ensure that the file can run. Add the following code to the app.py file underneath 
the app configuration and database setup:
\
`@app.route('/', methods=['POST', 'GET'])
def index():
	if request.method == 'POST':
		return redirect('/')
	else:
		return render_template('index.html')`
  
- For a much more in-depth explanation, visit this video by [freeCodeCamp.org](https://www.youtube.com/watch?v=Z1RJmh_OqeA)
- As for the models, My brother and I collected all the NHL Player data via [CapFriendly](https://www.capfriendly.com/), [HockeyDB](https://www.hockeydb.com/),
and [Hockey-Reference](https://www.hockey-reference.com/) and compiled them into several CSV files, present in the `Data` folder.
- The models were run through the `scikit-learn` library in python and the attributes deemed best to predict the contracts were selected and saved as .joblib files for later use
- The models used were Multiple Linear Regression Models, taking in up to 11 attributes and producing an output via Cap %
- We decided to use Cap % at signing of the contract instead of $ amount as the Cap changes year to year
- The [Predictor](Predictor/predictor.py) and [Interactions](Predictor/interactions.py) files were created and connect to each other
to help determine the contracts for each of the players
- The [Flask App](app.py) script had several methods added to it along with the required route decorators to ensure backend functionality. The user data was also processed and fed into the 
the `interaction.py` file to get the player valuations which would then be committed to the database
- The necessary HTML and CSS files (Although, Bootstrap 4 was used extensively in much of the program) were added, alongside some in-line JS to get the front-end working
	- Jinja 2 scripts were added inline as well to process the information received from the database and to ensure proper functionality of the code
- The app was then deployed to Heroku via GitHub

#### Test Run
- To Test the file, visit the [website](https://capredictor.herokuapp.com) or run it locally by running
the following code in your terminal:
\
`python app.py`

#### Deployment
- The program has been deployed on Heroku

### Authors
- The website was developed by myself, [Ali Raza Zaidi](https://github.com/AliRZ-02) and my brother, [Danyal Zaidi](https://github.com/DanyalZaidiPythonGithub)

### License
- This Software is licensed through the MIT License

### Learning Process 
- I learned quite a bit through this project, the likes of:
	- I learned how to create and deploy a flask app
	- I learned how to connect a front-end and back-end and how to use databases
	- I learned how to create data models in python using `scikit-learn`
	- I made myself more familiar with OOP Principles and their uses in Python
	- I became familiar with front-end frameworks like Bootstrap 4
	- I became familiar with Jinja2
	- I continued to get familiar with Git
	- I became more familiar with APIS and JSON Data
	- I became more familiar with automation through Python
	- I became more familiar with the data processing libraries in Python
	- I became more familiar with Heroku
	- I learned how to cohesively develop a full web application and deploy it

### Development
- This project is in its release phase now, although minor improvements are continually made

### Acknowledgements
- Visit the [Acknowledgements Page](acknowledgements.md)


CHANGELOG.md
# January 8 2021

- Updated File Comments
- Completed README
- Added CHANGELOG
- Added Acknowledgments
- Tweaked Flask App

Predictor/modeler.py
# Name: Ali Raza Zaidi
# Date: January 2, 2021
# Purpose: Create Models for future Contracts based on current player data

# Import Statements ---------------------------------------------------------------------------
import pandas
import joblib
from sklearn import linear_model
from matplotlib import pyplot as plt
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split
from scipy.stats import linregress

# Gathering Data ------------------------------------------------------------------------------
center_o27 = pandas.read_csv("..\Data\Clean\Centers-O27.csv")
center_u27_long = pandas.read_csv("..\Data\Clean\Centers-U27-Long.csv")
center_u27_short = pandas.read_csv("..\Data\Clean\Centers-U27-Short.csv")
defence_o27 = pandas.read_csv("..\Data\Clean\Defence-O27.csv")
defence_u27_short = pandas.read_csv("..\Data\Clean\Defence-U27-Short.csv")
defence_u27_long = pandas.read_csv("..\Data\Clean\Defence-U27-Long.csv")
goalies = pandas.read_csv("..\Data\Clean\Goalies.csv")
lw_o27 = pandas.read_csv("..\Data\Clean\LeftWings-O27.csv")
lw_u27_long = pandas.read_csv("..\Data\Clean\LeftWings-U27-Long.csv")
lw_u27_short = pandas.read_csv("..\Data\Clean\LeftWings-U27-Short.csv")
rw_o27 = pandas.read_csv("..\Data\Clean\RightWings-O27.csv")
rw_u27_long = pandas.read_csv("..\Data\Clean\RightWings-U27-Long.csv")
rw_u27_short = pandas.read_csv("..\Data\Clean\RightWings-U27-Short.csv")

# Data Splitting & Model Creation --------------------------------------------------------------

# UFA Centers
c_o27_in = center_o27[["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]]
c_o27_out = center_o27["Cap %"]

c_o27_reg = linear_model.LinearRegression()
c_o27_reg.fit(c_o27_in, c_o27_out)

# Non-Bridge RFA Centers
c_u27_long_in = center_u27_long[
    ["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]
]
c_u27_long_out = center_u27_long["Cap %"]

c_u27_long_reg = linear_model.LinearRegression()
c_u27_long_reg.fit(c_u27_long_in, c_u27_long_out)

# Bridge RFA Centers
c_u27_short_in = center_u27_short[
    ["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]
]
c_u27_short_out = center_u27_short["Cap %"]

c_u27_short_reg = linear_model.LinearRegression()
c_u27_short_reg.fit(c_u27_short_in, c_u27_short_out)

# UFA Defenceman
d_o27_in = defence_o27[
    [
        "Contract Length",
        "Age",
        "G/82",
        "A/82",
        "P/82",
        "P/GP",
        "Blocks/82",
        "Hits/82",
        "GP",
        "TOI/G",
        "S%",
    ]
]
d_o27_out = defence_o27["Cap %"]

d_o27_reg = linear_model.LinearRegression()
d_o27_reg.fit(d_o27_in, d_o27_out)

# Non-Bridge RFA Defenceman
d_u27_long_in = defence_u27_long[
    [
        "Contract Length",
        "Age",
        "G/82",
        "A/82",
        "P/82",
        "P/GP",
        "Blocks/82",
        "Hits/82",
        "GP",
        "TOI/G",
        "S%",
    ]
]
d_u27_long_out = defence_u27_long["Cap %"]

d_u27_long_reg = linear_model.LinearRegression()
d_u27_long_reg.fit(d_u27_long_in, d_u27_long_out)

# Bridge RFA Defenceman
d_u27_short_in = defence_u27_short[
    [
        "Contract Length",
        "Age",
        "G/82",
        "A/82",
        "P/82",
        "P/GP",
        "Blocks/82",
        "Hits/82",
        "GP",
        "TOI/G",
        "S%",
    ]
]
d_u27_short_out = defence_u27_short["Cap %"]

d_u27_short_reg = linear_model.LinearRegression()
d_u27_short_reg.fit(d_u27_short_in, d_u27_short_out)

# UFA Left-Wingers
lw_o27_in = lw_o27[["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]]
lw_o27_out = lw_o27["Cap %"]

lw_o27_reg = linear_model.LinearRegression()
lw_o27_reg.fit(lw_o27_in, lw_o27_out)

# Non-Bridge RFA LW
lw_u27_long_in = lw_u27_long[["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]]
lw_u27_long_out = lw_u27_long["Cap %"]

lw_u27_long_reg = linear_model.LinearRegression()
lw_u27_long_reg.fit(lw_u27_long_in, lw_u27_long_out)

# Bridge RFA LW
lw_u27_short_in = lw_u27_short[
    ["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]
]
lw_u27_short_out = lw_u27_short["Cap %"]

lw_u27_short_reg = linear_model.LinearRegression()
lw_u27_short_reg.fit(lw_u27_short_in, lw_u27_short_out)

# UFA RW
rw_o27_in = rw_o27[["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]]
rw_o27_out = rw_o27["Cap %"]

rw_o27_reg = linear_model.LinearRegression()
rw_o27_reg.fit(rw_o27_in, rw_o27_out)

# Non-Bridge RFA RW
rw_u27_long_in = rw_u27_long[["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]]
rw_u27_long_out = rw_u27_long["Cap %"]

rw_u27_long_reg = linear_model.LinearRegression()
rw_u27_long_reg.fit(rw_u27_long_in, rw_u27_long_out)

# Bridge RFA RW
rw_u27_short_in = rw_u27_short[
    ["Contract Length", "Age", "G/82", "A/82", "P/82", "P/GP"]
]
rw_u27_short_out = rw_u27_short["Cap %"]

rw_u27_short_reg = linear_model.LinearRegression()
rw_u27_short_reg.fit(rw_u27_short_in, rw_u27_short_out)

# Goalies
goalies_in = goalies[["Contract Length", "Games", "GAA", "SV%", "Win %"]]
goalies_out = goalies["Cap %"]

g_reg = linear_model.LinearRegression()
g_reg.fit(goalies_in, goalies_out)

# Model Dumping
joblib.dump(c_o27_reg, "..\Models\C-O27-Model.joblib")
joblib.dump(c_u27_long_reg, "..\Models\C-U27-L-Model.joblib")
joblib.dump(c_u27_short_reg, "..\Models\C-U27-S-Model.joblib")
joblib.dump(d_o27_reg, "..\Models\D-O27-Model.joblib")
joblib.dump(d_u27_long_reg, "..\Models\D-U27-L-Model.joblib")
joblib.dump(d_u27_short_reg, "..\Models\D-U27-S-Model.joblib")
joblib.dump(lw_o27_reg, "..\Models\L-O27-Model.joblib")
joblib.dump(lw_u27_long_reg, "..\Models\L-U27-L-Model.joblib")
joblib.dump(lw_u27_short_reg, "..\Models\L-U27-S-Model.joblib")
joblib.dump(rw_o27_reg, "..\Models\R-O27-Model.joblib")
joblib.dump(rw_u27_long_reg, "..\Models\R-U27-L-Model.joblib")
joblib.dump(rw_u27_short_reg, "..\Models\R-U27-S-Model.joblib")
joblib.dump(g_reg, "..\Models\G-Model.joblib")

Predictor/predictor.py
# Name: Ali Raza Zaidi
# Date: Jan 8 2021
# Purpose: Use Models created by modeler.py which exist in JOBLIB format and use them to predict player contracts based
# on given stats

# Import Statements
import joblib

# Main Class - Wanted to incorporate OOP Principles to make program more elegant and easier to use


class GetPrediction:
    def __init__(self, player, length, mode, age, resign, player_id=0):
        # Necessary Information to make a decent prediction
        self.player = player
        self.length = int(length)
        self.mode = mode
        self.age = age
        self.resign = resign
        self.player_id = player_id
        self.error = "An error occurred. Check if all inputs have been entered correctly and please try again"
        # Requirements for JOBLIB file
        if self.mode == 'G':
            self.group = ''
            self.bridge = ''
        elif self.age >= 27:
            self.group = 'O27-'
            self.bridge = ''
        elif self.length >= 4:
            self.group = 'U27-'
            self.bridge = 'L-'
        else:
            self.group = 'U27-'
            self.bridge = 'S-'

    def get_model(self):
        # Load the model for the specified position
        try:
            model = joblib.load('Models' + '/' + self.mode + '-' + self.group + self.bridge + 'Model.joblib')
            return model
        except:
            return FileNotFoundError

    def get_pred(self, valuation):
        # Use valuation received to send a personalized response for the given player
        current = ""

        for character in valuation:
            if character == '[':
                pass
            elif character == ']':
                pass
            else:
                current = current + character

        value = round((float(current) * 81500000))

        # NHL CBA dictates minimum contracts of 700K, so ny valuation less than that gets modeled as the league minimum
        if value <= 700000:
            value = 700000

        # NHL CBA Contract Length requirements have a max of 7 years for free agents while a max of 8 for resigning
        # players
        if not self.resign:
            self.length = int(self.length * (7/8))
        else:
            pass

        value = format(value, ",")  # Make the Valuation more readable with commas

        return str(self.player) + ":(2019-2020):$" + str(value)\
               + ":" + str(self.length) + ":" + str(self.player_id)

    def forward(self, g82, a82, p82, ppg):
        # Works with other methods to cohesively give a forward valuation
        try:
            model = self.get_model()
            prediction = model.predict([[self.length, self.age, g82, a82, p82, ppg]])
            return self.get_pred(str(prediction))
        except:
            return self.error + 'F'

    def defence(self, g82, a82, p82, ppg, b82, h82, gp, toi, spct):
        # Works with other methods to cohesively give a defence valuation
        try:
            model = self.get_model()
            prediction = model.predict([[self.length, self.age, g82, a82, p82, ppg, b82, h82, gp, toi, spct]])
            return self.get_pred(str(prediction))
        except:
            return self.error + 'D'

    def goalie(self, gp, gaa, svpct, winpct):
        # Works with other methods to cohesively give a goalie valuation
        try:
            model = self.get_model()
            prediction = model.predict([[self.length, gp, gaa, svpct, winpct]])
            return self.get_pred(str(prediction))
        except NameError:
            return self.error + 'G'


Helpers/player_images.py
# Name: Ali Raza Zaidi
# Date: Jan 8 2021
# Purpose: Download all player images from NHL API. Due to issues, with heroku, the player images weren't loading and so
#          I had to download all the player images locally and connect them to index.html that way

# Import Statements
import requests
import csv

# Getting all of the players from the idList CSV file
reader = csv.reader(open('../Data/Players/idList.csv', 'r', encoding="ISO-8859-1"))
players = {}
name_list = []

# Creating a Dictionary with player names and the corresponding player ID's to access the NHL API
for name, id in reader:
    if name != 'Names':
        players[name] = id
        name_list.append(name)

# Downloading the player images locally
for player in players:
    image = requests.get("http://nhl.bamcontent.com/images/headshots/current/168x168/"
                         + str(players[player]) + ".jpg")
    download = open(str(players[player])+".jpg", "wb")
    download.write(image.content)
    download.close()

Predictor/interactions.py
# Name: Ali Raza Zaidi
# Date: Jan 8 2021
# Purpose: Accesses the prediction models and returns the given prediction based
# on the input information and player position

# Import Statements
import csv
import requests
from Predictor import predictor
from difflib import get_close_matches

# Main Class - Wanted to incorporate OOP Principles to make program more elegant and easier to use


class GetInfo:
    def __init__(self, player_name, length):
        # All Methods will take in the Player Name and The Contract Length for their contract predictions
        self.player_name = player_name
        self.length = length

    def automatic(self, resign=True):
        # Automatic Mode uses the NHL API to find the value of the player's contract. It uses the player name
        # input from the user and finds the closest match amongst the list of 1090 active NHL player and uses
        # said player's id to access the NHL API and find their stats

        # Due to the way the NHL CBA is laid out, Free-Agents can only sign for a maximum of 7 Years with their
        # New teams while Re-signing players can sign for up to 8, hence why the adjustments have to be made
        if resign is False:
            self.length = self.length * (8/7)

        player_id = self.__id()

        # Tries to access the NHL API using the private method that allows for getting the closest ID Match
        # Alerts the user if no valid player ID is found
        try:
            player_stats = requests.get("https://statsapi.web.nhl.com/api/v1/people/" + str(player_id) +
                                       "/stats?stats=statsSingleSeason&season=20192020")
            player_data = requests.get("https://statsapi.web.nhl.com/api/v1/people/" + str(player_id))
        except:
                return "Such a player was not found. Try using manual mode"

        # Uses the JSON Data from the API to find the stats need to calculate the prediction
        try:
            position = player_data.json()['people'][0]["primaryPosition"]["code"]
            age = player_data.json()['people'][0]['currentAge']
            name = player_data.json()['people'][0]['fullName']
            stats = player_stats.json()['stats'][0]['splits'][0]['stat']
        except:
            return "Such a player was not found. Try using manual mode"

        # Using the Predictor Module, Sets up a GetPrediction object to find the player prediction and
        # calls the required module based on the given position
        get_pred = predictor.GetPrediction(name, self.length, position, age, resign, player_id)

        if position == 'L' or position == 'R' or position == 'C' or position == 'D':
            g82 = (stats['goals'] / stats['games']) * 82
            a82 = (stats['assists'] / stats['games']) * 82
            p82 = (stats['points'] / stats['games']) * 82
            ppg = (stats['points'] / stats['games'])
        else:
            pass

        # Finds GP Stat, which is needed for Goalie and Defence predictions
        gp = (stats['games'])
        if gp < 70:
            gp = (gp / 70) * 82
        else:
            gp = 82

        # Returns valuation for Forwards, Defencemen and Goalies based on Prediction model and stats from the API
        if position == 'L' or position == 'R' or position == 'C':
            valuation = get_pred.forward(g82,a82,p82,ppg)
            return valuation
        elif position == 'D':
            b82 = (stats['blocked'] / stats['games']) * 82
            h82 = (stats['hits'] / stats['games']) * 82
            toi_original = (stats['timeOnIcePerGame'])
            index = 0

            for char in toi_original:
                if char == ':':
                    break
                else:
                    index = index+1

            minutes = int(toi_original[:index])
            seconds = int(toi_original[index+1:]) / 60
            toi = minutes+seconds
            spct = int(stats['shotPct']) / 100
            valuation = get_pred.defence(g82, a82, p82, ppg, b82, h82, gp, toi, spct)
            return valuation
        elif position == 'G':
            gaa = (stats['goalAgainstAverage'])
            svpct = (stats['savePercentage'])
            winpct = (stats['wins'] / stats['games'])
            valuation = get_pred.goalie(gp, gaa, svpct, winpct)
            return valuation
        else:
            return "No player valuation could be found. Try again using manual mode"

    def forward(self, position, age, g82, a82, p82, ppg, resign=True):
        # For Manual Mode, uses input stats to find determine valuation from prediction model
        # Length changes made due to CBA restrictions
        if not resign:
            self.length = self.length * (8/7)

        algorithm = predictor.GetPrediction(self.player_name, self.length, position, age, resign)
        prediction = algorithm.forward(g82, a82, p82, ppg)
        return prediction

    def defence(self, age, g82, a82, p82, ppg, b82, h82, gp, toi, spct, resign=True):
        # For Manual Mode, uses input stats to find determine valuation from prediction model
        # Length changes made due to CBA restrictions
        if not resign:
            self.length = self.length * (8/7)
        algorithm = predictor.GetPrediction(self.player_name, self.length, 'D', age, resign)
        prediction = algorithm.defence(g82, a82, p82, ppg, b82, h82, gp, toi, spct)
        return prediction

    def goalie(self, age, gp, gaa, svpct, winpct, resign=True):
        # For Manual Mode, uses input stats to find determine valuation from prediction model
        # Length changes made due to CBA restrictions
        if not resign:
            self.length = self.length * (8/7)
        algorithm = predictor.GetPrediction(self.player_name, self.length, 'G', age, resign)
        prediction = algorithm.goalie(gp, gaa, svpct, winpct)
        return prediction

    def __id(self):
        # Uses the player's name and finds the corresponding Player ID given a CSV file containing both
        reader = csv.reader(open('Data/Players/idList.csv', 'r', encoding="ISO-8859-1"))
        players = {}
        name_list = []

        for name, id in reader:
            if name != 'Names':
                players[name] = id
                name_list.append(name)

        # Uses the difflib library to find the closest name match to the input name
        try:
            matches = get_close_matches(self.player_name, name_list, n=5)
            return players[matches[0]]
        except:
            return "No Player Found. Please try again using manual mode"

    def flag(self):
        # uses the player ID to find the nationality of the player to add their country's flag to the website
        player_id = self.__id()
        try:
            player_data = requests.get\
                ("https://statsapi.web.nhl.com/api/v1/people/"
                 + str(player_id))
        except:
            return "Such a player was not found. Try using manual mode"
        return player_data.json()['people'][0]['nationality']


app.py
# Name: Ali Raza Zaidi
# Date: Jan 8 2021
# Purpose: Flask App running the modeler in the backend

# Import Statements
import math
import sqlalchemy
from Predictor import interactions
from flask import Flask, render_template, request, redirect, flash
from datetime import datetime
from flask_sqlalchemy import SQLAlchemy

# Creating the Flask App and Configuring the Database through SQLAlchemy
app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI']='sqlite:///requests.db'
db = SQLAlchemy(app)

# Class that handles the database and the columns present in it
class Todo(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.String(200), nullable=False)
    photo = db.Column(db.String(200))
    nationality = db.Column(db.String(200))
    nationality_2 = db.Column(db.String(200))
    date_created = db.Column(db.DateTime, default=datetime.utcnow)

    def __repr__(self):
        return '<Task %r>' % self.id

# Main Website URL - Uses index.html


@app.route('/', methods=['POST', 'GET'])
def index():
    # Splits into POST and GET requests
    if request.method == 'POST':
        # Get information from submissions
        try:
            player_name = request.form['player_name']
            contract_length = round(float(request.form['length']))
            resign_statement = request.form['resign']
        except:
            return redirect('/')

        # Resign Statements are needed for contract length calculations
        if resign_statement == 'Re-sign':
            resign_statement = True
        else:
            resign_statement = False

        player_name = player_name.lower()  # Turn into lower-case to make name-matching easier

        # Test for Illegal Contract Length
        if contract_length <= 1:
            contract_length = 1
        elif contract_length >= 8 and resign_statement is True:
            contract_length = 8
        elif contract_length >= 7 and resign_statement is False:
            contract_length = 7
        else:
            pass

        # Find the Player Valuation
        money = interactions.GetInfo(player_name, contract_length)
        value = money.automatic(resign_statement)

        # If successful valuation is found
        if value != "No Player Found. Please try again using manual mode" \
                and value != "Such a player was not found. Try using manual mode":
            try:
                # Send info to db to put into the website
                id = value.split(":")[4]
                nations = {"CAN": "CA", "USA": "US", "SWE": "SE", "RUS": "RU",
                           "FIN": "FI", "CZE": "CZ", "CHE": "CH", "DNK": "DK",
                           "DEU": "DE", "SVK": "SK", "AUT": "AT", "SVN": "SI",
                           "FRA": "FR", "NOR": "NO", "LVA": "LV", "AUS": "AU",
                           "NLD": "NL"}
                nation = nations[money.flag()].lower()
                flag = "https://flagcdn.com/w160/" + nation + ".png"
                flag2 = "https://flagcdn.com/w320/" + nation + ".png 2x"

                # Put Gathered information into the DB
                new_player = Todo(name=value, photo=id, nationality=flag, nationality_2=flag2)

                # Remove Excess DB Elements
                reduce_db()
                check_duplicates(new_player)

                # Commit to DB
                db.session.add(new_player)
                db.session.commit()
                return redirect('/results')
            except:
                return "There was an error in searching the player. Please try again later"
        else:
            return redirect('/error')  # In the event of a player that couldn't be found

    else:
        players = Todo.query.order_by(Todo.date_created.desc()).all()
        return render_template('index.html', players= players)

# Once results are found, the results page, which is visually identical to the main page, is called.
# Likewise, in the case of an error, the error page is called


@app.route('/results', methods=['POST', 'GET'])
def results():
    return index()

@app.route('/error', methods=['POST', 'GET'])
def error():
    return index()

# Manual calls are directed to the manual endpoint


@app.route('/manual', methods=['GET'])
def manual():
    return render_template('manual.html')


@app.route('/manual/forward', methods=['POST', 'GET'])
def forward():
    # Split into POST and GET Requests
    if request.method == 'POST':
        # Get info from submission
        try:
            position = request.form['position']
            contract_length = round(float(request.form['length']))
            age = int(request.form['age'])
            resign_statement = request.form['resign']
            g82 = float(request.form['g82'])
            a82 = float(request.form['a82'])
            p82 = float(request.form['p82'])
        except:
            return redirect('/manual/forward')

        # Determine player position
        if position == 'RW':
            position = 'R'
        elif position == 'LW':
            position = 'L'
        else:
            pass

        # Resign status, needed for contract length calculations
        if resign_statement == 'Re-sign':
            resign_statement = True
        else:
            resign_statement = False

        # Checking for valid Contract Length
        if contract_length <= 1:
            contract_length = 1
        elif contract_length >= 8 and resign_statement is True:
            contract_length = 8
        elif contract_length >= 7 and resign_statement is False:
            contract_length = 7
        else:
            pass

        # Getting Valuation
        money = interactions.GetInfo(player_name="User-Generated-F", length= contract_length)
        value = money.forward(position, age, g82, a82, p82, ppg=(p82/82))

        # Putting gathered value into DB
        new_player = Todo(name=value)

        # Reducing Excess DB Elements
        reduce_db()
        check_duplicates(new_player)

        # Committing to DB
        try:
            db.session.add(new_player)
            db.session.commit()
            return redirect('/manual/forward/results')
        except:
            return "There was an error in searching the player. Please try again later"
    else:
        players = Todo.query.order_by(Todo.date_created.desc()).all()
        return render_template('forward.html', players= players)


@app.route('/manual/forward/results', methods=['POST', 'GET'])
def f_seen():
    return forward()


@app.route('/manual/defence', methods=['POST', 'GET'])
def defence():
    # Split into POST and GET requests
    if request.method == 'POST':
        # Get player information from submission
        try:
            gp = int(request.form['gp'])
            contract_length = round(float(request.form['length']))
            age = int(request.form['age'])
            resign_statement = request.form['resign']
            g82 = float(request.form['g82'])
            a82 = float(request.form['a82'])
            p82 = float(request.form['p82'])
            toi = float(request.form['toi'])
            toi_min = math.floor(toi)
            toi_sec = toi - toi_min
            toi = toi_min + ((toi_sec*100)/60)
            b82 = float(request.form['b82'])
            h82 = float(request.form['h82'])
            spct = float(request.form['spct'])
            spct = spct/100
        except:
            return redirect('/manual/defence')

        # Resign status needed for contract length calculations
        if resign_statement == 'Re-sign':
            resign_statement = True
        else:
            resign_statement = False

        # Checking for valid contract length
        if contract_length <= 1:
            contract_length = 1
        elif contract_length >= 8 and resign_statement is True:
            contract_length = 8
        elif contract_length >= 7 and resign_statement is False:
            contract_length = 7
        else:
            pass

        # Generating Valuation
        money = interactions.GetInfo("User-Generated-D", contract_length)
        value = money.defence(age, g82, a82, p82, (p82/82), b82, h82,
                              gp, toi, spct, resign_statement)

        # Adding valuation to DB
        new_player = Todo(name=value)

        # Reducing Excess DB Elements
        reduce_db()
        check_duplicates(new_player)

        # Committing to DB
        try:
            db.session.add(new_player)
            db.session.commit()
            return redirect('/manual/defence/results')
        except:
            return "There was an error in searching the player. Please try again later"
    else:
        players = Todo.query.order_by(Todo.date_created.desc()).all()
        return render_template('defence.html', players= players)


@app.route('/manual/defence/results', methods=['POST', 'GET'])
def d_seen():
    return defence()


@app.route('/manual/goalie', methods=['POST', 'GET'])
def goalie():
    # Split into POST and GET Requests
    if request.method == 'POST':
        # Get player information
        try:
            gp = int(request.form['gp'])
            contract_length = round(float(request.form['length']))
            age = int(request.form['age'])
            resign_statement = request.form['resign']
            gaa = float(request.form['gaa'])
            svpct = float(request.form['svpct'])
            winpct = float(request.form['winpct'])
        except:
            return redirect('/manual/defence')

        # Check player resign status for contract length calculations
        if resign_statement == 'Re-sign':
            resign_statement = True
        else:
            resign_statement = False

        # Checking for Valid Contract Length
        if contract_length <= 1:
            contract_length = 1
        elif contract_length >= 8 and resign_statement is True:
            contract_length = 8
        elif contract_length >= 7 and resign_statement is False:
            contract_length = 7
        else:
            pass

        # Generating Valuation
        money = interactions.GetInfo("User-Generated-G", contract_length)
        value = money.goalie(age, gp, gaa, svpct, winpct, resign_statement)

        # Adding valuation to DB
        new_player = Todo(name=value)

        # Reducing Excess DB Elements
        reduce_db()
        check_duplicates(new_player)

        # Committing to DB
        try:
            db.session.add(new_player)
            db.session.commit()
            return redirect('/manual/goalie/results')
        except:
            return "There was an error in searching the player. Please try again later"
    else:
        players = Todo.query.order_by(Todo.date_created.desc()).all()
        return render_template('goalie.html', players= players)


@app.route('/manual/goalie/results', methods=['POST', 'GET'])
def g_seen():
    return goalie()


def reduce_db():
    # Reduce DB Elements by collecting all current db elements and removing older queries or User-Generated queries
    players = Todo.query.order_by(Todo.date_created).all()
    if len(players) > 5:
        for player in players:
            if player.name == "User-Generated-F" or \
                player.name == "User-Generated-D" or \
                player.name == "User-Generated-G":
                    db.session.delete(player)
                    db.session.commit()
            else:
                pass
        players = Todo.query.order_by(Todo.date_created).all()
        if len(players) > 5:
            db.session.delete(players[0])
            db.session.commit()
    else:
        pass


def check_duplicates(content):
    # removes db element if another copy already exists
    players = Todo.query.order_by(Todo.date_created).all()
    for player in players:
        if content.name == player.name:
            db.session.delete(player)
            break
        else:
            pass


if __name__ == "__main__":
    app.run(debug=False)
